Documentation capturing enablement of NFD-Topology-Updater in NFD (#526)
kubernetes-sigs/node-feature-discovery: docs/get-started/deployment-and-usage.md

The template specs provided in the repo can be used directly:

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/manifests/nfd-master.yaml.template
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/manifests/nfd-worker-daemonset.yaml.template

> kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/manifests/nfd-topology-updater-daemonset.yaml.template

Reviewer: Update after kustomize rebase

Author: done

```

This will required RBAC rules and deploy nfd-master (as a deployment) and
nfd-worker and nfd-topology-updater (as daemonset) in the `node-feature-discovery` namespace.

Alternatively you can download the templates and customize the deployment
manually. For example, to deploy the [minimal](#minimal) image.

#### Master-worker pod

You can also run nfd-master, nfd-worker and nfd-topology-updater inside the same pod

```bash

> kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/manifests/nfd-daemonset-combined.yaml.template

Author: Ack, fixed

```bash
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/nfd-master.yaml.template
kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/nfd-worker-daemonset.yaml.template

> kubectl apply -f https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/nfd-topology-updater-daemonset.yaml.template

Author: Ack, will fix.

```bash
NUM_NODES=$(kubectl get no -o jsonpath='{.items[*].metadata.name}' | wc -w)

> curl -fs https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/manifests/nfd-worker-job.yaml.template | \

Reviewer: Update after kustomize rebase

    kubectl apply -f -
```

The example above launches as many jobs as there are non-master nodes. Note that
this approach does not guarantee running once on every node. For example,
tainted, non-ready nodes or some other reasons in Job scheduling may cause some
node(s) will run extra job instance(s) to satisfy the request.

#### Topology Updater One-shot

NFD Topology Updater can alternatively be configured as a one-shot job.
The Job template may be used to achieve this:

```bash
NUM_NODES=$(kubectl get no -o jsonpath='{.items[*].metadata.name}' | wc -w)

> curl -fs https://raw.githubusercontent.com/kubernetes-sigs/node-feature-discovery/{{ site.release }}/nfd-topology-updater-job.yaml.template | \

Author: Good catch, I will update the manifests in the implementation PR.

#### Master-worker pod

> You can also run nfd-master, nfd-worker and nfd-topology-updater inside the same pod

Reviewer 2: This isn't true, topology-updater is not deployed by the `default-combined` overlay. You could mention that. And possibly add that in order to enable topology-updater in this scenario the user must customize the deployment herself.

```bash
kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/default-combined?ref={{ site.release }}

```

> This creates a DaemonSet runs nfd-worker, nfd-master and nfd-topology-updater in the same Pod.

Reviewer 2: The above comment applies here, too

this approach does not guarantee running once on every node. For example,
tainted, non-ready nodes or some other reasons in Job scheduling may cause some
node(s) will run extra job instance(s) to satisfy the request.

#### Topologyupdater pod

NFD Topology Updater can be configured along with the default overlay, all
deployed as separate pods. The `topologyupdater` overlay may be used along
with `default` overlay to achieve this:

Deploy the `default` overlay:

```bash
kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/default?ref={{ site.release }}
```

Deploy the `topologyupdater` overlay:

```bash
kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/master-worker-topologyupdater?ref={{ site.release }}
```

#### Master Worker Topologyupdater

NFD Master, NFD worker and NFD Topology Updater can be configured to be deployed
as separate pods. The `master-worker-topologyupdater` overlay may be used to
achieve this:

```bash
kubectl apply -k https://github.com/kubernetes-sigs/node-feature-discovery/deployment/overlays/master-worker-topologyupdater?ref={{ site.release }}

```

> #### Topology Updater One-shot

Reviewer 2: nit: only capitalize the first word of section heading (might apply elsewhere, too...)

```suggestion
#### Topologyupdater one-shot
```

NFD-Worker is preferably run as a Kubernetes DaemonSet. This assures
re-labeling on regular intervals capturing changes in the system configuration
and makes sure that new nodes are labeled as they are added to the cluster.
Worker connects to the nfd-master service to advertise hardware features.

When run as a daemonset, nodes are re-labeled at an interval specified using
the `-sleep-interval` option. In the
[template](https://github.com/kubernetes-sigs/node-feature-discovery/blob/{{site.release}}/manifests/nfd-worker-daemonset.yaml.template#L26)
the default interval is set to 60s which is also the default when no
`-sleep-interval` is specified. Also, the configuration file is re-read on
each iteration providing a simple mechanism of run-time reconfiguration.

### NFD-Topology-Updater

NFD-Worker is preferably run as a Kubernetes DaemonSet. This assures

> re-examination (and CRD updation) on regular intervals capturing changes in the

Reviewer 2: ```suggestion
re-examination (and CR update) on regular intervals capturing changes in the
```

Author: Ack

allocated resources and hence the allocatable resources on a per zone basis. It

> makes sure that more CRD instances are created as new nodes get added to the

Reviewer 2: ```suggestion
makes sure that more CR instances are created as new nodes get added to the
```

Author: Ack

> cluster. Topology-Updater connects to the nfd-master service to creates CRD

Reviewer 3: to create

Reviewer 2: CRD -> CR

provides a simple mechanism of dynamic run-time reconfiguration. See
[worker configuration](#worker-configuration) for more details.

### NFD-Topology-Updater

NFD-Topology-Updater is preferably run as a Kubernetes DaemonSet. This assures

> re-examination (and CR updation) on regular intervals capturing changes in

Reviewer 2: ```suggestion
re-examination (and CR updates) on regular intervals capturing changes in
```

the allocated resources and hence the allocatable resources on a per zone

> basis. It makes sure that more CR instances are created as new nodes get

Reviewer 2: Suggestion: add a reference to the *NodeResourceTopology CR* section of `introduction.md`
